### PR TITLE
fix: resolve warnings and improve file extraction logic

### PR DESCRIPTION
**Pull request by ShigrafS**
Closes: #755 

**Description**

This draft PR addresses and aims to fix all existing errors, warnings, and lint issues throughout the codebase to improve overall code quality, stability, and maintainability. It includes fixes in utilities, sanitization, and other core packages with the following key changes:

* Fixed improper resource handling such as missing or misused `defer` statements (e.g., closing files and readers).

This PR serves as a foundation for a cleaner, more maintainable codebase by systematically resolving technical debt and improving developer experience.

**Notes for Reviewers**

* This is a draft PR and may require iterative reviews and tests.
* Some changes are large refactors; please pay close attention to file handling and error wrapping.
* Unit tests and integration tests should be run to ensure no regressions.
* Feedback on potential edge cases or improvements in error messaging is welcome.
* Additional lint rules might be applied progressively in subsequent PRs.

**[[Signed commits](https://chatgpt.com/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**

* [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by vr-varad** [comment]
Thank you for your contribution!!
If this work is complete, this is a good item to add to the weekly [Meshery Development meeting agenda](https://docs.google.com/document/d/15IEwtTxPOkQXN2r_1LzGXuAv2mIUlmbdtmxpyxXF-KI/edit). You can add this item in the doc, attend, and present it. Meeting details at https://meet.layer5.io/meshery.

**Comment by ShigrafS** [comment]
Hi @vr-varad 
Actually the issue that this PR is trying to solve has a varied scope. 
Therefore, I'm planning to reduce the scope of this PR and fix the original issue with multiple PRs. 
Additionally, there are more changes that needs to be pushed to this branch, after I decide on my approach. 

Please let me know your opinions. 
Thanks!

**Comment by lekaf974** [comment]
@ShigrafS I strongly suggest you create several PRs and fix issues 1 by 1 please, this will help reviewers

File: utils/utils.go [hunk 1]
```diff
@@ -145,13 +145,15 @@ func CreateFile(contents []byte, filename string, location string) error {
 		return err
 	}
 
+	// Write contents to the file
```
Comment (lekaf974) [inline utils/utils.go]: This comment is not useful

File: utils/utils.go [hunk 2]
```diff
@@ -145,13 +145,15 @@ func CreateFile(contents []byte, filename string, location string) error {
...
 	if _, err = fd.Write(contents); err != nil {
-		fd.Close()
+		fd.Close() // Close the file before returning the error
```
Comment (lekaf974) [inline utils/utils.go]: Comment has been added but should we not che k if it returns an error ?

File: utils/utils.go [hunk 3]
```diff
@@ -145,13 +145,15 @@ func CreateFile(contents []byte, filename string, location string) error {
...
 		return err
 	}
 
-	if err = fd.Close(); err != nil {
-		return err
+	// Check error while closing the file
+	if err := fd.Close(); err != nil {
+		return fmt.Errorf("failed to close file descriptor: %w", err)
```
Comment (lekaf974) [inline utils/utils.go]: Should return a meshkit error

File: utils/unarchive.go [hunk 4]
```diff
@@ -149,22 +150,28 @@ func ExtractTarGz(path, downloadfilePath string) error {
...
 			}
 			if _, err = io.Copy(outFile, tarReader); err != nil {
+				outFile.Close()
 				return ErrExtractTarXZ(err, path)
 			}
-			outFile.Close()
+
+			// Close the file and check for errors
+			if err := outFile.Close(); err != nil {
+				return fmt.Errorf("failed to close output file %s: %w", header.Name, err)
```
Comment (lekaf974) [inline utils/unarchive.go]: Should return meskit error

File: utils/unarchive.go [hunk 5]
```diff
@@ -149,22 +150,28 @@ func ExtractTarGz(path, downloadfilePath string) error {
...
 		case tar.TypeReg:
+			// Ensure that the directory for the file exists
 			_ = os.MkdirAll(filepath.Join(path, filepath.Dir(header.Name)), 0755)
 			var outFile *os.File
 			outFile, err = os.Create(filepath.Join(path, header.Name))
 			if err != nil {
 				return ErrExtractTarXZ(err, path)
 			}
 			if _, err = io.Copy(outFile, tarReader); err != nil {
+				outFile.Close()
```
Comment (lekaf974) [inline utils/unarchive.go]: Should we not handle error here ?

File: utils/walker/git.go [hunk 6]
```diff
@@ -144,7 +144,9 @@ func clonewalk(g *Git) error {
 	var wg sync.WaitGroup
 	defer func() {
 		wg.Wait()
-		os.RemoveAll(path)
+		if err := os.RemoveAll(path); err != nil {
+			fmt.Println(fmt.Errorf("failed to remove path %s: %w", path, err))
```
Comment (lekaf974) [inline utils/walker/git.go]: We ahould use meshkit logger

File: generators/github/git_repo.go [hunk 7]
```diff
@@ -47,9 +47,15 @@ func (gr GitRepo) GetContent() (models.Package, error) {
 	br := bufio.NewWriter(fd)
 
 	defer func() {
-		br.Flush()
-		fd.Close()
+		if err := br.Flush(); err != nil {
+			fmt.Printf("Warning: failed to flush writer: %v\n", err)
```
Comment (lekaf974) [inline generators/github/git_repo.go]: Should use meshkit logger

**Review by lekaf974 — CHANGES_REQUESTED** [review]
Could you adress comments

File: utils/unarchive.go [hunk 8]
```diff
@@ -106,31 +105,30 @@ func ExtractZip(path, artifactPath string) error {
...
 	return nil
-
 }
 
 func ExtractTarGz(path, downloadfilePath string) error {
 	gzipStream, err := os.Open(downloadfilePath)
 	if err != nil {
 		return ErrReadFile(err, downloadfilePath)
 	}
+	defer gzipStream.Close()
```
Comment (lekaf974) [inline utils/unarchive.go]: why do we not handling error ?

File: utils/unarchive.go [hunk 9]
```diff
@@ -158,23 +156,35 @@ func ExtractTarGz(path, downloadfilePath string) error {
...
 
 			// Close the file and check for errors
 			if err := outFile.Close(); err != nil {
-				return fmt.Errorf("failed to close output file %s: %w", header.Name, err)
+				return ErrFileClose(err, header.Name)
 			}
 
 		default:
-			return ErrExtractTarXZ(err, path)
+			return ErrUnsupportedTarHeaderType(header.Typeflag, header.Name)
```
Comment (lekaf974) [inline utils/unarchive.go]: Is this error the most generic that it could happens ?

File: utils/unarchive.go [hunk 10]
```diff
@@ -158,23 +156,35 @@ func ExtractTarGz(path, downloadfilePath string) error {
...
 		}
 	}
 
 	return nil
 }
 
+// ErrFileClose creates a standardized error for file closing failures.
+func ErrFileClose(err error, filename string) error {
```
Comment (lekaf974) [inline utils/unarchive.go]: this error is not a valid meshkit error also there is one already existing https://github.com/meshery/meshkit/blob/3c002e7a2ca7aaea8c6f4ed63f6781ed7a4d0025/utils/error.go#L250

**Review by lekaf974 — CHANGES_REQUESTED** [review]
@ShigrafS just made some comments

File: utils/unarchive.go [hunk 11]
```diff
@@ -149,25 +148,43 @@ func ExtractTarGz(path, downloadfilePath string) error {
...
+	return fmt.Errorf("failed to close file %s: %w", filename, err)
+}
+
+// ErrUnsupportedTarHeaderType creates an error for unsupported tar header types.
+func ErrUnsupportedTarHeaderType(typeflag byte, name string) error {
```
Comment (lekaf974) [inline utils/unarchive.go]: this is not a valid meshkit error

File: utils/unarchive.go (hunk 8 above)
Comment (ShigrafS) [reply utils/unarchive.go]: The error from `openedFile.Close()1` is intentionally ignored in this case because the error from `CopyBuffer` is considered more critical.

File: utils/unarchive.go (hunk 9 above)
Comment (ShigrafS) [reply utils/unarchive.go]: This error could be more generic. We could create a more general error type for issues related to file extraction and copying (e.g., `ErrFileExtraction`), which would allow us to catch errors from different operations in the same category.

File: utils/kubernetes/kompose/convert.go [hunk 12]
```diff
@@ -42,52 +43,93 @@ func Convert(dockerCompose DockerComposeFile) (string, error) {
...
+	if err := os.Remove(resultFilePath); err != nil {
+		return fmt.Errorf("failed to remove result file %s: %w", resultFilePath, err)
+	}
+
+	return nil // No errors
+}
+
+// formatComposeFile takes in a pointer to the compose file byte array and formats it
+// so that it is compatible with Kompose. It expects a validated Docker Compose file.
+func formatComposeFile(yamlManifest *DockerComposeFile) error {
```
Comment (lekaf974) [inline utils/kubernetes/kompose/convert.go:110]: Have we evaluate the impact on services calling this function since the signature has changed ?

File: utils/walker/git.go [hunk 13]
```diff
@@ -175,7 +182,7 @@ func clonewalk(g *Git) error {
 	}
 
 	if !info.IsDir() {
-		err = g.readFile(info, rootPath)
+		err = g.readFile(info, rootPath, log)
```
Comment (lekaf974) [inline utils/walker/git.go:185]: why adding logger, here ?

**Comment by lekaf974** [comment]
@ShigrafS sorry for dealy and thanks for the work done, added small comments

File: utils/kubernetes/kompose/convert.go [hunk 14]
```diff
@@ -42,52 +43,93 @@ func Convert(dockerCompose DockerComposeFile) (string, error) {
...
+		return "", cleanupErr
+	}
+
 	return string(result), nil
 }
 
-type composeFile struct {
-	Version string `yaml:"version,omitempty"`
+// cleanup removes temporary files
+func cleanup(tempFilePath, resultFilePath string) error {
```
Comment (lekaf974) [inline utils/kubernetes/kompose/convert.go]: I don't see the advantage to extract this in a dedicate function since it is bind to those files.

Also it should be done in a defer like previous implementation to ensure the files are closed properly. Here if one of the previous action fail the files will never be closed

**Review by lekaf974 — CHANGES_REQUESTED** [review]
changes to ensure files are closed properly

File: utils/kubernetes/kompose/convert.go (hunk 12 above)
Comment (ShigrafS) [reply utils/kubernetes/kompose/convert.go:110]: `formatComposeFile` is only called once from within the `Convert` function in this same file.

File: utils/kubernetes/kompose/convert.go [hunk 15]
```diff
@@ -42,37 +43,55 @@ func Convert(dockerCompose DockerComposeFile) (string, error) {
 	tempFilePath := filepath.Join(mesheryDir, "temp.data")
 	resultFilePath := filepath.Join(mesheryDir, "result.yaml")
 
+	// Defer cleanup to ensure temp files are removed even on error.
+	defer func() {
+		// We log cleanup errors but do not return them, as the primary function error is more important.
+		if err := os.Remove(tempFilePath); err != nil && !os.IsNotExist(err) {
+			// A logger should be used here to warn about the failed cleanup.
+			// e.g., log.Warnf("failed to remove temporary file %s: %v", tempFilePath, err)
```
Comment (lekaf974) [inline utils/kubernetes/kompose/convert.go]: Why not log here ?

File: utils/kubernetes/kompose/convert.go [hunk 16]
```diff
@@ -42,37 +43,55 @@ func Convert(dockerCompose DockerComposeFile) (string, error) {
...
+		}
+		if err := os.Remove(resultFilePath); err != nil && !os.IsNotExist(err) {
+			// e.g., log.Warnf("failed to remove result file %s: %v", resultFilePath, err)
```
Comment (lekaf974) [inline utils/kubernetes/kompose/convert.go]: same

File: utils/kubernetes/kompose/convert.go [hunk 17]
```diff
@@ -42,37 +43,55 @@ func Convert(dockerCompose DockerComposeFile) (string, error) {
 	tempFilePath := filepath.Join(mesheryDir, "temp.data")
 	resultFilePath := filepath.Join(mesheryDir, "result.yaml")
 
+	// Defer cleanup to ensure temp files are removed even on error.
```
Comment (lekaf974) [inline utils/kubernetes/kompose/convert.go]: not util IMO

File: utils/kubernetes/kompose/convert.go (hunk 14 above)
Comment (lekaf974) [reply utils/kubernetes/kompose/convert.go]: No more used, write ?

File: utils/kubernetes/kompose/convert.go (hunk 12 above)
Comment (lekaf974) [reply utils/kubernetes/kompose/convert.go:110]: you're right, I did not notice it was a local function. Sorry for the noise

**Review by lekaf974 — CHANGES_REQUESTED** [review]
Small comments added

File: utils/kubernetes/kompose/convert.go (hunk 14 above)
Comment (ShigrafS) [reply utils/kubernetes/kompose/convert.go]: Yes. This function is no longer used since the logic was moved into the `defer` block for safer cleanup. I've removed the dead code.

**Comment by ShigrafS** [comment]
@lekaf974 Apologies for the delay. 
I've made the necessary changes. 
Kindly review it. 
Thank you.
